Implement Change Password Feature

Habeeb4shittu opened the pull request:
This PR adds the Change Password functionality, allowing authenticated users to securely update their account passwords.

Key Updates:

Added changePassword method implement

- [x] ation in PasswordService to handle password validation and update.
- [x] Integrated controller endpoint for handling password change requests.
- [x] Added form validation for current_password, new_password, and confirm_password.
- [x] Improved error handling and success response structure via ApiResponse.
- [x] Ensured password update is restricted to authenticated users.

Testing Steps:

1. Send a POST request to /api/password/change with:
    - current_password
    - new_password
    - new_password_confirmation

2. Verify password is updated only if current password is correct.

3. Confirm appropriate error messages for invalid credentials or unauthenticated users.

Status: ✅ Ready for Review

Habeeb4shittu committed:
feat: implement change password functionality with validation

mysticCoder100 commented on app/Http/Controllers/Password/PasswordController.php:
```diff
@@ -51,10 +51,37 @@ public function resetPassword(ResetPasswordRequest $request)
+        $user = $this->getLoggedInUser();
+        if (!$user) {
+            return ApiResponse::error(message: 'User is not authenticated.', status: 401);
+        }
```
Use auth:sanctum middleware in the route as this route is meant for authenticated users

mysticCoder100 commented on app/Http/Controllers/Password/PasswordController.php:
```diff
@@ -51,10 +51,37 @@ public function resetPassword(ResetPasswordRequest $request)
+            'new_password' => 'required|string|min:8|confirmed',
+        ]);
+
+        self::logInfo('Password Data ', [
```
Don't log this, it isn't necessary

mysticCoder100 commented on routes/password/password-routes.php:
```diff
@@ -11,4 +11,7 @@
     Route::post('/reset/{email}/{token}', [PasswordController::class, 'resetPassword']);
+    Route::post('/change', [PasswordController::class, 'changePassword'])
+        ->middleware('auth:sanctum')
+        ->middleware('throttle:5,1');
```
you already use the auth:sanctum here no need checking if the user is logged in, in the controller. this throttle is not good ooo 5 attempts in a minute, let make it like that of reset password

mysticCoder100 commented on app/Services/Password/PasswordService.php:
```diff
@@ -92,4 +92,31 @@ public function resetPassword(ResetPasswordDto $dto): bool
             throw new PasswordServiceException('Unable to reset the password');
         }
     }
+    public function changePassword(int $userId, string $currentPassword, string $newPassword): bool
```
you don't need to pass the user ID here, the hasAuthenticatedUser trait is used in this class, just use it and update it on it. do you understand

mysticCoder100 reviewed (changes requested):
I dropped some inline comments, kindly look through

Habeeb4shittu committed:
Code cleanups and fixes

mysticCoder100 committed:
feat: implement change password functionality

This commit adds a new change password feature allowing authenticated users to update
their passwords securely. Changes include:

Core Features:
- Add change password endpoint under `/api/password/change/{user}`
- Implement password change validation and security checks
- Add refresh token cleanup after password change

Authentication & Security:
- Require authentication via sanctum middleware
- Validate current password before allowing changes
- Implement rate limiting (8 attempts per 10 minutes)
- Clear all access tokens and refresh token after password change

Data Transfer:
- Add ChangePasswordDto for handling password change requests
- Update PasswordServiceContract and PasswordService implementations
- Add password validation rules (minimum length, confirmation)

Tests:
- Add comprehensive test suite for password change functionality
- Test authentication requirements
- Test validation rules
- Test security constraints
- Test session management

Additional:
- Update route configuration for password management
- Add proper error handling and responses
- Add audit logging for password changes
